mysql/resource_grant.go

```diff
@@ -485,168 +593,311 @@ func ImportGrant(ctx context.Context, d *schema.ResourceData, meta interface{})
-	d.SetId(id)
+	if tableGrant, ok := grant.(*TablePrivilegeGrant); ok {
+		d.Set("database", tableGrant.Database)
+		d.Set("grant", grant.GrantOption())
+		d.Set("tls_option", tableGrant.TLSOption)
 
-	d.Set("user", user)
-	d.Set("host", host)
-	d.Set("database", database)
-	d.Set("table", grant.Table)
-	d.Set("grant", grant.Grant)
-	d.Set("tls_option", "NONE")
-	d.Set("privileges", grant.Privileges)
+	} else if procedureGrant, ok := grant.(*ProcedurePrivilegeGrant); ok {
+		d.Set("database", fmt.Sprintf("%s %s.%s", procedureGrant.ObjectT, procedureGrant.Database, procedureGrant.CallableName))
+		d.Set("table", "*")
+		d.Set("grant", grant.GrantOption())
+		d.Set("tls_option", procedureGrant.TLSOption)
+
+	} else if roleGrant, ok := grant.(*RoleGrant); ok {
+		d.Set("grant", grant.GrantOption())
+		d.Set("roles", roleGrant.Roles)
+		d.Set("tls_option", roleGrant.TLSOption)
+	} else {
+		panic("Unknown grant type")
+	}
+
+	// Only set privileges if there is a delta in the normalized privileges
+	if grantWithPriv, hasPriv := grant.(MySQLGrantWithPrivileges); hasPriv {
+		currentPriv, ok := d.GetOk("privileges")
+		if !ok {
+			d.Set("privileges", grantWithPriv.GetPrivileges())
+		} else {
+			currentPrivs := setToArray(currentPriv.(*schema.Set))
+			currentPrivs = normalizePerms(currentPrivs)
+			if !reflect.DeepEqual(currentPrivs, grantWithPriv.GetPrivileges()) {
+				d.Set("privileges", grantWithPriv.GetPrivileges())
+			}
+		}
+	}
+
+	// This is a bit of a hack, since we don't have a way to distingush between users and roles
+	// from the grant itself. We can only infer it from the schema.
+	userOrRole := grant.GetUserOrRole()
+	if d.Get("role") != "" {
+		d.Set("role", userOrRole.Name)
+	} else {
+		d.Set("user", userOrRole.Name)
+		d.Set("host", userOrRole.Host)
+	}
 
 	return d
 }
 
-func showGrant(ctx context.Context, db *sql.DB, user, database, table string, grantOption bool) (MySQLGrant, error) {
-	allGrants, err := showUserGrants(ctx, db, user)
+func hasConflictingGrants(ctx context.Context, db *sql.DB, desiredGrant MySQLGrant) (MySQLGrant, error) {
+	allGrants, err := showUserGrants(ctx, db, desiredGrant.GetUserOrRole())
 	if err != nil {
-		return MySQLGrant{}, fmt.Errorf("showGrant - getting all grants failed: %w", err)
-	}
-	grants := MySQLGrant{
-		Database: database,
-		Table:    table,
-		Grant:    grantOption,
-	}
-	for _, grant := range allGrants {
-		// We must normalize database as it may contain something like PROCEDURE `asd` or the same without backticks.
-		// TODO: write tests or consider some other way to handle permissions to PROCEDURE/FUNCTION
-		if grant.Grant == grantOption {
-			if normalizeDatabase(grant.Database) == normalizeDatabase(database) && grant.Table == table {
-				grants.Privileges = append(grants.Privileges, grant.Privileges...)
+		return nil, fmt.Errorf("showGrant - getting all grants failed: %w", err)
+	}
+	for _, dbGrant := range allGrants {
+		if desiredGrant.GrantOption() != dbGrant.GrantOption() {
```

Comment:
I believe grant option difference still makes two grants conflicting - or do you believe otherwise?

Reply:
I believe they have to be non-conflicting. Suppose I wanted to give `UserX` the ability to `SELECT`, and grant that `SELECT` privilege to other users. However, I also wanted to give `UserX` the ability to `INSERT` without the ability to grant that to other users. The only way to express that in MySQL syntax is with two distinct `GRANT` statements.

Reply:
I see your point, but it doesn't work.
Hopefully this example with Percona 8.0 (fork of MySQL 8.0) shows where it fails - you can easily reproduce it in Docker:
```
mysql> create user 'experiment'@'%';
Query OK, 0 rows affected (0,03 sec)

mysql> show grants for 'experiment'@'%';
+----------------------------------------+
| Grants for experiment@%                |
+----------------------------------------+
| GRANT USAGE ON *.* TO `experiment`@`%` |
+----------------------------------------+
1 row in set (0,00 sec)

mysql> create database `testdb`;
Query OK, 1 row affected (0,03 sec)

mysql> grant insert on testdb.* to 'experiment'@'%';
Query OK, 0 rows affected (0,02 sec)

mysql> show grants for 'experiment'@'%';
+------------------------------------------------+
| Grants for experiment@%                        |
+------------------------------------------------+
| GRANT USAGE ON *.* TO `experiment`@`%`         |
| GRANT INSERT ON `testdb`.* TO `experiment`@`%` |
+------------------------------------------------+
2 rows in set (0,00 sec)

mysql> grant select on testdb.* to 'experiment'@'%' with grant option;
Query OK, 0 rows affected (0,02 sec)

mysql> show grants for 'experiment'@'%';
+--------------------------------------------------------------------------+
| Grants for experiment@%                                                  |
+--------------------------------------------------------------------------+
| GRANT USAGE ON *.* TO `experiment`@`%`                                   |
| GRANT SELECT, INSERT ON `testdb`.* TO `experiment`@`%` WITH GRANT OPTION |
+--------------------------------------------------------------------------+
2 rows in set (0,00 sec)
```
See the last grants.

Reply:
Good catch! I would have expected two separate rows.  I will remove this condition.